File: Code/Mantid/Framework/PythonInterface/plugins/algorithms/CreateTransmissionWorkspaceAuto.py
"""*WIKI* 

Facade over [[CreateTransmissionWorkspace]]. Pull numeric parameters out of the instrument parameters where possible. You can override any of these automatically
applied defaults by providing your own value for the input.

See [[CreateTransmissionWorkspace]] for more information on the wrapped algorithm.

*WIKI*"""


import sys
from mantid.simpleapi import CreateTransmissionWorkspace
from mantid.api import *
from mantid.kernel import *

class CreateTransmissionWorkspaceAuto(DataProcessorAlgorithm):
    
    def category(self):
        return "Reflectometry\\ISIS"

    def name(self):
        return "CreateTransmissionWorkspaceAuto"
    
    def PyInit(self):
    
        self.setOptionalMessage("Creates a transmission run workspace in Wavelength from input TOF workspaces.")
        self.setWikiSummary("Creates a transmission run workspace in Wavelength from input TOF workspaces. See [[Reflectometry_Guide]]")
        
        analysis_modes = ["PointDetectorAnalysis", "MultiDetectorAnalysis"]
        analysis_mode_validator = StringListValidator(analysis_modes)
        
        self.declareProperty(name="AnalysisMode", defaultValue=analysis_modes[0], validator=analysis_mode_validator, direction = Direction.Input, doc="Analysis Mode to Choose")
        
        input_validator = WorkspaceUnitValidator("TOF")
        self.declareProperty(MatrixWorkspaceProperty("FirstTransmissionRun", "", Direction.Input, validator=input_validator), "Input workspace")
        self.declareProperty(MatrixWorkspaceProperty("SecondTransmissionRun", "", optional=PropertyMode.Optional,  validator=input_validator, direction=Direction.Input), "Second transmission run workspace in TOF")
        self.declareProperty(MatrixWorkspaceProperty("OutputWorkspace", "", Direction.Output), "Output transmission workspace in wavelength")

        self.declareProperty(FloatArrayProperty(name="Params", values=[sys.maxint]), doc="Rebinning Parameters. See Rebin for format.")
        self.declareProperty(name="StartOverlap", defaultValue=sys.float_info.max, doc="Overlap in Q.")
        self.declareProperty(name="EndOverlap", defaultValue=sys.float_info.max, doc="End overlap in Q.")
        index_bounds = IntBoundedValidator()
        index_bounds.setLower(0)
        self.declareProperty(name="I0MonitorIndex", defaultValue=sys.maxint, validator=index_bounds, doc="I0 monitor index" )
        self.declareProperty(name="ProcessingInstructions", direction=Direction.Input, defaultValue="", doc="Processing instructions to extract spectra belonging to detector workspace, see [[PerformIndexOperations]] for syntax.")
        self.declareProperty(name="WavelengthMin", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Wavelength Min in angstroms")
        self.declareProperty(name="WavelengthMax", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Wavelength Max in angstroms")
        self.declareProperty(name="WavelengthStep", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Wavelength step in angstroms")
        self.declareProperty(name="MonitorBackgroundWavelengthMin", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Monitor wavelength background min in angstroms")
        self.declareProperty(name="MonitorBackgroundWavelengthMax", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Monitor wavelength background max in angstroms")
        self.declareProperty(name="MonitorIntegrationWavelengthMin", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Monitor integral min in angstroms")
        self.declareProperty(name="MonitorIntegrationWavelengthMax", direction=Direction.Input, defaultValue=sys.float_info.max, doc="Monitor integral max in angstroms")
        
        
    def value_or_none(self, prop_name, instrument):
        property = self.getProperty(prop_name)
        if property.isDefault:
            return None # TODO: check this
        else:
            return property.value
    
        
    def value_to_apply(self, prop_name, instrument, idf_name=""):
        property = self.getProperty(prop_name)
        if property.isDefault:
            return instrument.getNumberParameter(idf_name)[0] 
        else:
            return property.value
        
    def PyExec(self):
        first_ws = self.getProperty("FirstTransmissionRun").value
        instrument = first_ws.getInstrument()
        
        '''
        Get all the inputs.
        '''
        outputWorkspaceName = self.getPropertyValue("OutputWorkspace")
        
        analysis_mode = self.getPropertyValue("AnalysisMode")
        
        second_ws_property = self.getProperty("SecondTransmissionRun")
        second_ws = None
        if not second_ws_property.isDefault:
            second_ws = second_ws_property.value
        
        start_overlap = self.value_or_none("StartOverlap", instrument)
        
        end_overlap = self.value_or_none("EndOverlap", instrument)
        
        params = self.value_or_none("Params", instrument)
        
        i0_monitor_index = int(self.value_to_apply("I0MonitorIndex", instrument, "I0MonitorIndex"))
        
        processing_commands = str()
        workspace_index_list = list()
        if self.getProperty("ProcessingInstructions").isDefault:
            if analysis_mode == "PointDetectorAnalysis":
                workspace_index_list.append( int(instrument.getNumberParameter("PointDetectorStart")[0]) )
                workspace_index_list.append( int(instrument.getNumberParameter("PointDetectorStop")[0]) )
            else:
                workspace_index_list.append( int(instrument.getNumberParameter("MultiDetectorStart")[0]) )
                workspace_index_list.append( first_ws.getNumberHistograms() - 1)
            processing_commands = ','.join(map(str, workspace_index_list))
        else:
            processing_commands = self.getProperty("ProcessingInstructions").value
        
        wavelength_min = self.value_to_apply("WavelengthMin", instrument, "LambdaMin")
        
        wavelength_max = self.value_to_apply("WavelengthMax", instrument, "LambdaMax")
        
        wavelength_step = self.value_or_none("WavelengthStep", instrument)
        
        wavelength_back_min = self.value_to_apply("MonitorBackgroundWavelengthMin", instrument, "MonitorBackgroundMin")
        
        wavelength_back_max = self.value_to_apply("MonitorBackgroundWavelengthMax", instrument, "MonitorBackgroundMax")
        
        wavelength_integration_min = self.value_to_apply("MonitorIntegrationWavelengthMin", instrument, "MonitorIntegralMin")
        
        wavelength_integration_max = self.value_to_apply("MonitorIntegrationWavelengthMax", instrument, "MonitorIntegralMax")
        
        '''
        Pass the arguments and execute the main algorithm.
        '''
        output_ws = CreateTransmissionWorkspace(FirstTransmissionRun=first_ws, SecondTransmissionRun=second_ws, OutputWorkspace=outputWorkspaceName, 
                                    StartOverlap=start_overlap, EndOverlap=end_overlap, Params=params, I0MonitorIndex=i0_monitor_index,
                                    ProcessingInstructions=processing_commands, WavelengthMin=wavelength_min, WavelengthStep=wavelength_step, WavelengthMax=wavelength_max,
                                    MonitorBackgroundWavelengthMin=wavelength_back_min, MonitorBackgroundWavelengthMax=wavelength_back_max,
                                    MonitorIntegrationWavelengthMin=wavelength_integration_min, MonitorIntegrationWavelengthMax=wavelength_integration_max)
                                    
        
        self.setProperty("OutputWorkspace", output_ws)
    
AlgorithmFactory.subscribe(CreateTransmissionWorkspaceAuto)

File: Code/Mantid/Framework/Algorithms/src/CreateTransmissionWorkspaceAuto.cpp
/*WIKI*
 Facade over [[CreateTransmissionWorkspace]]. Pull numeric parameters out of the instrument parameters where possible. You can override any of these automatically
 applied defaults by providing your own value for the input.

 See [[CreateTransmissionWorkspace]] for more information on the wrapped algorithm.
 *WIKI*/

#include "MantidAlgorithms/CreateTransmissionWorkspaceAuto.h"
#include "MantidKernel/RebinParamsValidator.h"
#include "MantidAPI/WorkspaceValidators.h"
#include "MantidKernel/ListValidator.h"
#include "MantidKernel/ArrayProperty.h"
#include "MantidKernel/BoundedValidator.h"
#include "MantidAPI/AlgorithmManager.h"

using namespace Mantid::Kernel;
using namespace Mantid::API;

namespace Mantid
{
  namespace Algorithms
  {

    // Register the algorithm into the AlgorithmFactory
    DECLARE_ALGORITHM(CreateTransmissionWorkspaceAuto)

    //----------------------------------------------------------------------------------------------
    /** Constructor
     */
    CreateTransmissionWorkspaceAuto::CreateTransmissionWorkspaceAuto()
    {
    }

    //----------------------------------------------------------------------------------------------
    /** Destructor
     */
    CreateTransmissionWorkspaceAuto::~CreateTransmissionWorkspaceAuto()
    {
    }

    //----------------------------------------------------------------------------------------------
    /// Sets documentation strings for this algorithm
    const std::string CreateTransmissionWorkspaceAuto::summary() const
    {
      return "Creates a transmission run workspace in Wavelength from input TOF workspaces.";
    }

    //----------------------------------------------------------------------------------------------
    /** Initialize the algorithm's properties.
     */
    void CreateTransmissionWorkspaceAuto::init()
    {

      std::vector<std::string> analysis_modes;
      analysis_modes.push_back("PointDetectorAnalysis");
      analysis_modes.push_back("MultiDetectorAnalysis");
      declareProperty("AnalysisMode", analysis_modes.at(0),
          boost::make_shared<StringListValidator>(analysis_modes), "Analysis Mode to Choose",
          Direction::Input);

      declareProperty(
          new WorkspaceProperty<MatrixWorkspace>("FirstTransmissionRun", "", Direction::Input,
              boost::make_shared<WorkspaceUnitValidator>("TOF")), "Input workspace.");
      declareProperty(
          new WorkspaceProperty<MatrixWorkspace>("SecondTransmissionRun", "", Direction::Input,
              PropertyMode::Optional, boost::make_shared<WorkspaceUnitValidator>("TOF")),
          "Second transmission run workspace in TOF.");
      declareProperty(new WorkspaceProperty<MatrixWorkspace>("OutputWorkspace", "", Direction::Output),
          "Output transmission workspace in wavelength.");

      declareProperty(
          new ArrayProperty<double>("Params", boost::make_shared<RebinParamsValidator>(true)),
          "A comma separated list of first bin boundary, width, last bin boundary. "
              "These parameters are used for stitching together transmission runs. "
              "Values are in wavelength (angstroms). This input is only needed if a SecondTransmission run is provided.");

      declareProperty(
          new PropertyWithValue<double>("StartOverlap", Mantid::EMPTY_DBL(), Direction::Input),
          "Start wavelength for stitching transmission runs together");

      declareProperty(new PropertyWithValue<double>("EndOverlap", Mantid::EMPTY_DBL(), Direction::Input),
          "End wavelength (angstroms) for stitching transmission runs together");

      auto boundedIndex = boost::make_shared<BoundedValidator<int> >();
      boundedIndex->setLower(0);

      declareProperty(new PropertyWithValue<int>("I0MonitorIndex", Mantid::EMPTY_INT(), boundedIndex),
          "I0 monitor index");

      declareProperty(new PropertyWithValue<std::string>("ProcessingInstructions", "", Direction::Input),
          "Processing instructions on workspace indexes to yield only the detectors of interest. See [[PerformIndexOperations]] for details.");

      declareProperty("WavelengthMin", Mantid::EMPTY_DBL(), "Wavelength Min in angstroms",
          Direction::Input);
      declareProperty("WavelengthMax", Mantid::EMPTY_DBL(), "Wavelength Max in angstroms",
          Direction::Input);
      declareProperty("WavelengthStep", Mantid::EMPTY_DBL(), "Wavelength step in angstroms",
          Direction::Input);
      declareProperty("MonitorBackgroundWavelengthMin", Mantid::EMPTY_DBL(),
          "Monitor wavelength background min in angstroms", Direction::Input);
      declareProperty("MonitorBackgroundWavelengthMax", Mantid::EMPTY_DBL(),
          "Monitor wavelength background max in angstroms", Direction::Input);
      declareProperty("MonitorIntegrationWavelengthMin", Mantid::EMPTY_DBL(),
          "Monitor integral min in angstroms", Direction::Input);
      declareProperty("MonitorIntegrationWavelengthMax", Mantid::EMPTY_DBL(),
          "Monitor integral max in angstroms", Direction::Input);
    }

    //----------------------------------------------------------------------------------------------
    /** Execute the algorithm.
     */
    void CreateTransmissionWorkspaceAuto::exec()
    {
      //auto firstWS = AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(this->getPointerToProperty("FirstTransmissionRun")->value());
      MatrixWorkspace_sptr firstWS = this->getProperty("FirstTransmissionRun");
      auto instrument = firstWS->getInstrument();

      //Get all the inputs.

      std::string outputWorkspaceName = this->getPropertyValue("OutputWorkspace");
      std::string analysis_mode = this->getPropertyValue("AnalysisMode");

      MatrixWorkspace_sptr secondWS = this->getProperty("SecondTransmissionRun");

      auto start_overlap = isSet<double>("StartOverlap");
      auto end_overlap = isSet<double>("EndOverlap");
      auto params = isSet<std::vector<double>>("Params");
      auto i0_monitor_index = static_cast<int>(checkForDefault("I0MonitorIndex", instrument,
          "I0MonitorIndex"));

      std::string processing_commands;
      if (this->getPointerToProperty("ProcessingInstructions")->isDefault())
      {
        const int start = static_cast<int>(instrument->getNumberParameter("PointDetectorStart")[0]);
        const int stop = static_cast<int>(instrument->getNumberParameter("PointDetectorStop")[0]);
        if (analysis_mode == "PointDetectorAnalysis")
        {
          if (start == stop)
          {
            processing_commands = boost::lexical_cast<std::string>(start);
          }
          else
          {
            processing_commands = boost::lexical_cast<std::string>(start) + ","
                + boost::lexical_cast<std::string>(stop);
          }
        }
        else
        {
          processing_commands = boost::lexical_cast<std::string>(
              static_cast<int>(instrument->getNumberParameter("MultiDetectorStart")[0])) + ","
              + boost::lexical_cast<std::string>(firstWS->getNumberHistograms() - 1);
        }
      }
      else
      {
        std::string processing_commands_temp = this->getProperty("ProcessingInstructions");
        processing_commands = processing_commands_temp;
      }

      double wavelength_min = checkForDefault("WavelengthMin", instrument, "LambdaMin");
      double wavelength_max = checkForDefault("WavelengthMax", instrument, "LambdaMax");
      auto wavelength_step = isSet<double>("WavelengthStep");
      double wavelength_back_min = checkForDefault("MonitorBackgroundWavelengthMin", instrument,
          "MonitorBackgroundMin");
      double wavelength_back_max = checkForDefault("MonitorBackgroundWavelengthMax", instrument,
          "MonitorBackgroundMax");
      double wavelength_integration_min = checkForDefault("MonitorIntegrationWavelengthMin", instrument,
          "MonitorIntegralMin");
      double wavelength_integration_max = checkForDefault("MonitorIntegrationWavelengthMax", instrument,
          "MonitorIntegralMax");

      //construct the algorithm

      IAlgorithm_sptr algCreateTransWS = createChildAlgorithm("CreateTransmissionWorkspace");
      algCreateTransWS->setRethrows(true);
      algCreateTransWS->initialize();

      if (algCreateTransWS->isInitialized())
      {

        algCreateTransWS->setProperty("FirstTransmissionRun", firstWS);

        if (secondWS)
        {
          algCreateTransWS->setProperty("SecondTransmissionRun", secondWS);
        }

        algCreateTransWS->setProperty("OutputWorkspace", outputWorkspaceName);

        if (start_overlap.is_initialized())
        {
          algCreateTransWS->setProperty("StartOverlap", start_overlap.get());
        }
        if (end_overlap.is_initialized())
        {
          algCreateTransWS->setProperty("EndOverlap", end_overlap.get());
        }
        if (params.is_initialized())
        {
          algCreateTransWS->setProperty("Params", params.get());
        }

        algCreateTransWS->setProperty("I0MonitorIndex", i0_monitor_index);
        algCreateTransWS->setProperty("ProcessingInstructions", processing_commands);
        algCreateTransWS->setProperty("WavelengthMin", wavelength_min);

        if (wavelength_step.is_initialized())
        {
          algCreateTransWS->setProperty("WavelengthStep", wavelength_step.get());
        }

        algCreateTransWS->setProperty("WavelengthMax", wavelength_max);
        algCreateTransWS->setProperty("MonitorBackgroundWavelengthMin", wavelength_back_min);
        algCreateTransWS->setProperty("MonitorBackgroundWavelengthMax", wavelength_back_max);
        algCreateTransWS->setProperty("MonitorIntegrationWavelengthMin", wavelength_integration_min);
        algCreateTransWS->setProperty("MonitorIntegrationWavelengthMax", wavelength_integration_max);

        algCreateTransWS->execute();
        if (!algCreateTransWS->isExecuted())
        {
          throw std::runtime_error("CreateTransmissionWorkspace did not execute sucessfully");
        }
        else
        {
          MatrixWorkspace_sptr outWS = algCreateTransWS->getProperty("OutputWorkspace");
          setProperty("OutputWorkspace", outWS);
        }
      }
      else
      {
        throw std::runtime_error("CreateTransmissionWorkspace could not be initialised");
      }

    }

    template<typename T>
    boost::optional<T> CreateTransmissionWorkspaceAuto::isSet(std::string propName) const
    {
      auto algProperty = this->getPointerToProperty(propName);
      if (algProperty->isDefault())
      {
        return boost::optional<T>();
      }
      else
      {
        T value = this->getProperty(propName);
        return boost::optional<T>(value);
      }
    }

    double CreateTransmissionWorkspaceAuto::checkForDefault(std::string propName,
        Mantid::Geometry::Instrument_const_sptr instrument, std::string idf_name) const
    {
      auto algProperty = this->getPointerToProperty(propName);
      if (algProperty->isDefault())
      {
        auto defaults = instrument->getNumberParameter(idf_name);
        if (defaults.size() == 0)
        {
          throw std::runtime_error(
              "No data could be retrieved from the parameters and argument wasn't provided: "
                  + propName);
        }
        return defaults[0];
      }
      else
      {
        return boost::lexical_cast<double, std::string>(algProperty->value());
      }
    }

  } // namespace Algorithms
} // namespace Mantid
